src/abstractdataframe/join.jl

```diff
@@ -337,10 +358,27 @@ function Base.join(df1::AbstractDataFrame,
                 push!(leftonly_ixs, l_ix)
             end
         end
-        return joiner.dfl[leftonly_ixs, :]
+        joined = joiner.dfl[leftonly_ixs, :]
     else
         throw(ArgumentError("Unknown kind of join requested: $kind"))
     end
+
+    if !(indicator == nothing)
+        left = joined[indicator_cols[1]]
+        left = Int.(coalesce.(left, false))
+
+        right = joined[indicator_cols[2]]
+        right = Int.(coalesce.(right, false)) * 2
+
+        _merge = DataFrame(indicator => categorical(left + right))
+        _merge[indicator] = recode(_merge[indicator], 1 => "left_only", 2 => "right_only", 3 => "both")
```

Comment:
You can make all of this more efficient:
- instead of allocating the result in `left` and `right`, loop over both columns at the same time in a comprehension and allocate directly the final vector `refs`, with a `UInt8` eltype
- use the low-level `CategoricalArray` constructor `CategoricalArray{String,1}(refs, CategoricalPool{String,UInt8}(["left_only", "right_only", "both"]))`
- only create the `DataFrame` at that point

Reply:
I'm afraid I can't get this to work -- when I try the low-level construction I get: 

```
CategoricalArray{String,1,UInt8}(refs, CategoricalPool{String,UInt8}(["left_only", "right_only", "both"]))
ERROR: MethodError: no method matching CategoricalArray{String,1,UInt8,V,C,U} where U where C where V(::Array{Int64,1}, ::CategoricalPool{String,UInt8,CategoricalString{UInt8}})
Closest candidates are:
  CategoricalArray{String,1,UInt8,V,C,U} where U where C where V(::AbstractArray; ordered) where {T, N, R} at /Users/Nick/.julia/packages/CategoricalArrays/AxkM/src/array.jl:170
  CategoricalArray{String,1,UInt8,V,C,U} where U where C where V(::Any) where T<:AbstractArray at abstractarray.jl:22
```

And I can't seem to find documentation for the kind of construction you're suggesting...

Reply:
Yeah it's hard to find the right combination of parameters which works (these are low-level, so not for general consumption and not really documented). But have you tried the exact one I suggested? In my tests, it worked.

Reply:
Looking at the code in the PR, the problem is that you should do `refs = UInt8[...]`